File: src/Algorithm/Recursion/Leetcode_Problems/Palindrome_Partitioning/Palindrome_Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s)
    {
        vector<vector<string>> ans;
        vector<string> sub;
        partition(s, 0, sub, ans);
        return ans;
    }
    
    void partition(string& s, int i, vector<string>& sub, vector<vector<string>>& ans)
    {   
        int l = s.length();
    
        if(i == l)
        {
            ans.push_back(sub);
            return;
        }
    
        for(int j = i; j < l; ++j)
        {
            string temp = s.substr(i, j - i + 1);
            if(isPalindrome(temp))
            {
                sub.push_back(temp);
                partition(s, j + 1, sub, ans);
                sub.pop_back();
            }
        }
    }

    
    bool isPalindrome(string s)
    {
        int l = s.length();
    
        for(int i = 0; i < l / 2; ++i)
            if(s[i] != s[l - 1 - i])
                return false;
    
        return true;
    }
};

File: src/Algorithm/Recursion/Leetcode_Problems/Palindrome_Partitioning/PalindromePartitioning.java
package Algorithm.Recursion.Leetcode_Problems.Palindrome_Partitioning;

import java.util.ArrayList;
import java.util.List;

public class PalindromePartitioning {
    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();
        List<String> now = new ArrayList<>();
        helper(0, s, now, ans);
        return ans;
    }
    public void helper(int index, String s, List<String> curr, List<List<String>> ans){
        if(index == s.length()){
            ans.add(new ArrayList<>(curr));
            return;
        }
        for(int i = index; i < s.length(); i++){
            if(isPalindrome(s, index, i)){
                curr.add(s.substring(index, i + 1));
                helper(i + 1, s, curr, ans);
                curr.remove(curr.size() - 1); 
            }
        }
    } 
    public boolean isPalindrome(String s, int l, int r){ 
        while(l <= r){
            if(s.charAt(l++) != s.charAt(r--)) return false;
        }
        return true;
    }
}

// Time Complexity: O(N*2^N) 
// Space Complexity: O(N)